### PR TITLE
Fixes #25792: String.split(Char) stops delegating to Regex

### LINKED ISSUE #25792: StringOps.split(Char) brings in the whole Regex machinery
I have noticed a large JS bundle (Scala.JS) increase when doing a simple split by a character, see the example below.

Then I have seen that split by Char implementation just delegates to a Regex:

```scala
def split(separator: Char): Array[String] = s.split(escape(separator))
```

Purely because this specialized method exists, I assumed it's implemented in a more efficient way. Is it worth doing better and save some CPU cycles and JS bytes?

## Minimized example

```Scala
"a/b/c".split('/')
```

### PR DESCRIPTION
**Pull request by strelec**
Fixes #25792

## How much have you relied on LLM-based tools in this contribution?

Not at all

## How was the solution tested?

Covered by existing tests (this is a refactoring)

Plus I ran the below fuzzing code:

```scala
	@main
	def splitTestFuzzing: Unit = {
		val cases = {
			val low = 0xDF62.toChar
			val high = 0xD852.toChar
			Array(40.toChar, low, high)
		}
		val alphabet =
			cases.flatMap(ch => Seq(ch, (ch + 1).toChar))

		for {
			_ <- 1 to 2000000
			len = Random.nextInt(8)
			s = Array.fill(len)(alphabet(Random.nextInt(6))).mkString
			ch = cases(Random.nextInt(3))
		} // assert it matches existing split function s.split(ch)
	}
```

### CODE REVIEW

**Comment by strelec** [comment]
The code is written in a way that 99.99% case is performant, this is when the split char is not a surrogate.

File: library/src/scala/collection/StringOps.scala [hunk 1]
```diff
@@ -898,7 +898,53 @@ final class StringOps(private val s: String) extends AnyVal { self =>
    *
    *  @return an array of strings computed by splitting this string around occurrences of the separator character
    */
-  def split(separator: Char): Array[String] = s.split(escape(separator))
+  def split(separator: Char): Array[String] = {
+    var end = s.length
+    if (end == 0) return Array("")
+
```
Comment (tanishiking) [inline library/src/scala/collection/StringOps.scala:919]: Optimization: fast path for no match

```scala
    if (s.indexOf(separator) < 0) return Array(s)
```

**Review by tanishiking — COMMENTED** [review]
Thanks, I roughly benchmarked the new impl, and it is faster for most cases (except no match case)!
It seems there's no unit tests for `split(char: Char)` in `library/test/scala/collection/StringOpsTest.scala`, can you add test for the semantics?

FYI @sjrd

**Comment by tanishiking** [comment]
@scala/stdlib-officers

File: library/src/scala/collection/StringOps.scala (hunk 1 above)
Comment (strelec) [reply library/src/scala/collection/StringOps.scala:919]: Interesting, this is a good idea, because no-match case is probably very common in the wild.

But you only benefit from `indexOf` call when it is `< 0`, otherwise you discard returned index.

Would it make sense to capture the return value, then use it for the first recursive call on lines 913 and 928?

**Comment by strelec** [comment]
> Thanks, I roughly benchmarked the new impl, and it is faster for most cases (except no match case)! It seems there's no unit tests for `split(char: Char)` in `library/test/scala/collection/StringOpsTest.scala`, can you add test for the semantics?
> 
> FYI @sjrd

I will add tests, and revamp the scaladoc.

Even though code speedup is not the primary motivation of this CL, what are the % improvements like, approximately?

File: library/src/scala/collection/StringOps.scala (hunk 1 above)
Comment (sjrd) [reply library/src/scala/collection/StringOps.scala:919]: That's going to make every other case slower, though. I expect `split` to actually split something on virtually every call. We shouldn't make the rare case fast by slowing down the common case.

Comment (strelec) [reply library/src/scala/collection/StringOps.scala:919]: Which is mitigated if we capture `s.indexOf(separator)` and reuse it in place of the first recursive call.

However, maybe it's not worth bothering if the no-match case does not occur often in the wild...

**Comment by tanishiking** [comment]
I tested with the following benchmark, and for extreme case (many separator), it's about 1.5 times faster than before. For normal case like only one separator there, it was basically same as before.

<details>
<summary>benchmark</summary>

```bash
$ scala-cli  run --power --jmh bench.scala
```

```scala
//> using scala 3.8.3

package bench

import java.lang.Character
import java.util.concurrent.TimeUnit

import scala.compiletime.uninitialized

import org.openjdk.jmh.annotations.*

@BenchmarkMode(Array(Mode.AverageTime))
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Warmup(iterations = 3, time = 500, timeUnit = TimeUnit.MILLISECONDS)
@Measurement(iterations = 5, time = 500, timeUnit = TimeUnit.MILLISECONDS)
@Fork(1)
@State(Scope.Thread)
class StringSplitBench:
  @Param(Array(
    "no-sep",
    "one-sep",
    "many-seps",
  ))
  var benchCase: String = uninitialized

  private var input: String = uninitialized
  private var separator: Char = uninitialized

  @Setup
  def setup(): Unit =
    benchCase match
      case "no-sep" =>
        input = "abcdefghijklmnopqrstuvwxyz0123456789" * 16
        separator = '|'
      case "one-sep" =>
        input = "abcdefghijklmnopqrstuvwxyz.0123456789"
        separator = '.'
      case "many-seps" =>
        input = (0 until 64).mkString(",")
        separator = ','

  @Benchmark
  def oldRegex(): Array[String] =
    oldSplit(input, separator)

  @Benchmark
  def newPr(): Array[String] =
    newSplit(input, separator)

  private def escape(ch: Char): String =
    if ch >= 'a' && ch <= 'z' || ch >= 'A' && ch <= 'Z' || ch >= '0' && ch <= '9' then ch.toString
    else "\\" + ch

  private def oldSplit(s: String, separator: Char): Array[String] =
    s.split(escape(separator))

  private def newSplit(s: String, separator: Char): Array[String] =
    var end = s.length
    if end == 0 then return Array("")

    while end != 0 && s.charAt(end - 1) == separator do end -= 1
    if end == 0 then return Array.empty

    val builder = Array.newBuilder[String]

    if !Character.isSurrogate(separator) then
      @annotation.tailrec
      def collect(start: Int): Unit =
        s.indexOf(separator, start) match
          case -1 =>
            builder += s.substring(start)
          case idx =>
            builder += s.substring(start, idx)
            if idx != end then collect(idx + 1)
      collect(0)
    else
      val isLowSurrogate = Character.isLowSurrogate(separator)
      if isLowSurrogate && end != s.length && Character.isHighSurrogate(s.charAt(end - 1)) then end += 1

      @annotation.tailrec
      def collect(start: Int, searchFrom: Int): Unit =
        s.indexOf(separator, searchFrom) match
          case -1 =>
            builder += s.substring(start)
          case idx =>
            if idx == end then
              builder += s.substring(start, idx)
            else if isLowSurrogate && idx != 0 && Character.isHighSurrogate(s.charAt(idx - 1)) then
              collect(start, idx + 1)
            else if !isLowSurrogate && s.charAt(idx + 1).isLowSurrogate then
              collect(start, idx + 2)
            else
              builder += s.substring(start, idx)
              collect(idx + 1, idx + 1)
      collect(0, 0)

    builder.result()
```

</details>

**Comment by He-Pin** [comment]
## Review: Performance & Correctness Analysis

**Correctness: Verified safe.** After tracing through the surrogate handling logic and multiple edge cases, the implementation is correct:

- The trailing separator trim (`while end != 0 && s(end - 1) == separator`) ensures bounds safety — if `indexOf` finds a separator at the trimmed position, `idx == end` triggers early return, preventing any out-of-bounds access.
- Surrogate pair validation correctly skips matches that are part of valid high-low pairs.
- All edge cases verified: empty string, all separators, trailing surrogates, bare surrogates, and well-formed surrogate pairs.

**Performance suggestion (optional but recommended):**

The benchmark shows the new impl is slightly slower than regex for the "no match" case. A simple early exit would fix this:

```scala
val firstIdx = s.indexOf(separator)
if (firstIdx == -1) return Array(s)
```

This avoids the trailing trim loop, builder allocation, and all surrogate logic when the separator doesn't appear at all — likely the most common case in practice.

Looking forward to the tests you mentioned adding!

File: library/src/scala/collection/StringOps.scala [hunk 2]
```diff
@@ -898,7 +898,53 @@ final class StringOps(private val s: String) extends AnyVal { self =>
    *
    *  @return an array of strings computed by splitting this string around occurrences of the separator character
    */
-  def split(separator: Char): Array[String] = s.split(escape(separator))
+  def split(separator: Char): Array[String] = {
+    var end = s.length
+    if (end == 0) return Array("")
```
Comment (He-Pin) [inline library/src/scala/collection/StringOps.scala]: a cached Array("")

Comment (sjrd) [reply library/src/scala/collection/StringOps.scala]: No. You can't do that. `Array` is mutable. You cannot give a cached mutable instance and give it to the outside world.

File: library/src/scala/collection/StringOps.scala [hunk 3]
```diff
@@ -898,7 +898,53 @@ final class StringOps(private val s: String) extends AnyVal { self =>
...
+
+    while (end != 0 && s(end - 1) == separator) end -= 1
+    if (end == 0) return Array.empty
```
Comment (sjrd) [inline library/src/scala/collection/StringOps.scala]: This might be just me, but I find it hard to read all those condition side effects at the end of lines. Consider formatting things so that side effects are always on their own line. For example here:
```scala
    while (end != 0 && s(end - 1) == separator)
      end -= 1
    if (end == 0)
      return Array.empty
```

File: library/src/scala/collection/StringOps.scala [hunk 4]
```diff
@@ -898,7 +898,53 @@ final class StringOps(private val s: String) extends AnyVal { self =>
...
+
+    val builder = Array.newBuilder[String]
+
+    if (!separator.isSurrogate) {
+      @annotation.tailrec
+      def collect(start: Int): Unit =
```
Comment (sjrd) [inline library/src/scala/collection/StringOps.scala]: Any reason for an `@tailrec` method rather than a good old `while` loop? tailrec methods are great when the code is actually functional inside. But here we have a tailrec method that performs side effects by appending to `builder`. A `while` loop is probably easier to follow.

**Review by sjrd — COMMENTED** [review]
The algorithm looks good to me. But yes, we should have dedicated tests for this method. Previously we relied on the fact the implementation was straightforward, but that's not the case anymore.

File: library/src/scala/collection/StringOps.scala (hunk 3 above)
Comment (som-snytt) [reply library/src/scala/collection/StringOps.scala]: It used to be just you, but you persuaded me previously that it's nicer not to bury side effects in our peripheral vision.

**Comment by som-snytt** [comment]
>  I expect split to actually split something on virtually every call.

> when the separator doesn't appear at all — likely the most common case in practice.

I agree with sjrd's expectation, but it should incur the cost only when the first separator is seen.

I object to this use of em dash — there's no contrasting emphasis, just a comma splice with different punctuation. (Edit: that was joking indignation.)

File: library/src/scala/collection/StringOps.scala (hunk 3 above)
Comment (strelec) [reply library/src/scala/collection/StringOps.scala]: Done.

File: library/src/scala/collection/StringOps.scala (hunk 4 above)
Comment (strelec) [reply library/src/scala/collection/StringOps.scala]: Every `var` adds a cognitive load, so it better be justified. I personally find a recursive version so much easier to understand, seeing that it terminates, and generally check the corectness.

This is assuming that we trust the compiler to produce good code when using @tailrec.

File: library/src/scala/collection/StringOps.scala (hunk 2 above)
Comment (strelec) [reply library/src/scala/collection/StringOps.scala]: Possibly we could do it for Array.empty to avoid an allocation (Unless JDK optimizes it)

**Comment by strelec** [comment]
Two commits added:

1. Removed one if-statement from the main codepath, performance should improve across the board.
2. Added a test.

**Comment by som-snytt** [comment]
I pushed a commit for doc and a commit to practice writing a `while do ()`. It might be approximately correct.

I can't tell by looking whether it's easier to read. I did not jmh.

There is a convenience method `assertSameElements` to fix the test.

`builder.result()` wants parens.

Edit: casually tried the jmh. It fixes the no-sep case, with a slight 20% improvement in one-sep.

**Comment by strelec** [comment]
I find the imperative version less understandable, less obviously correct. It is also longer, and it mutates two indices (`i` and `start`), while before only `start` was mutated (and even then, only indirectly by parameter passing). Also It does not look nice if one branch is recursive and the other not, you lose the feeling that they are actually the same code with a bit added on top.

I added one proposed optimization - Do indexOf at the start of the method, rather than later. It's a small commit, check it.

My proposal - let's commit my original work, and split off the discussion / potential improvements into another pull request.

**Comment by som-snytt** [comment]
> it should incur the cost only when the first separator is seen.

The solution should include the `no-sep` regression.

I don't disagree with your description, but the motivation for less-readable imperative loop is the regression, that is, besides the sjrd challenge.

Feel free to rewrite to recursive and ignore my commit. My commit was only POC proof of concept.

I don't mind a style difference between the branches, since the else is a different use case (to put it strongly).

I didn't give any thought to the benchmark, which might be misleading. The mere numbers I saw were as below. I don't know why the `one-sep` case is lower.

<details>

```
3.8.3
Benchmark               (benchCase)  Mode  Cnt     Score    Error  Units
StringSplitBench.newPr       no-sep  avgt    5    15.017 ±  2.864  ns/op
StringSplitBench.newPr      one-sep  avgt    5    59.941 ±  1.475  ns/op
StringSplitBench.newPr    many-seps  avgt    5  1177.811 ± 53.198  ns/op

Benchmark               (benchCase)  Mode  Cnt     Score    Error  Units
StringSplitBench.newPr       no-sep  avgt    5    14.816 ±  2.340  ns/op
StringSplitBench.newPr      one-sep  avgt    5    59.039 ±  8.640  ns/op
StringSplitBench.newPr    many-seps  avgt    5  1105.268 ± 26.919  ns/op

3.9
Benchmark               (benchCase)  Mode  Cnt     Score     Error  Units
StringSplitBench.newPr       no-sep  avgt    5    16.810 ±   3.071  ns/op
StringSplitBench.newPr      one-sep  avgt    5    41.739 ±   0.794  ns/op
StringSplitBench.newPr    many-seps  avgt    5  1218.067 ± 459.005  ns/op

Benchmark               (benchCase)  Mode  Cnt     Score    Error  Units
StringSplitBench.newPr       no-sep  avgt    5    15.401 ±  1.360  ns/op
StringSplitBench.newPr      one-sep  avgt    5    40.946 ±  0.701  ns/op
StringSplitBench.newPr    many-seps  avgt    5  1164.987 ± 79.203  ns/op

PR
Benchmark               (benchCase)  Mode  Cnt     Score    Error  Units
StringSplitBench.newPr       no-sep  avgt    5    39.990 ±  0.675  ns/op
StringSplitBench.newPr      one-sep  avgt    5    57.109 ±  2.204  ns/op
StringSplitBench.newPr    many-seps  avgt    5  1194.925 ± 53.903  ns/op

Benchmark               (benchCase)  Mode  Cnt     Score    Error  Units
StringSplitBench.newPr       no-sep  avgt    5    25.916 ±  0.269  ns/op
StringSplitBench.newPr      one-sep  avgt    5    54.880 ±  3.803  ns/op
StringSplitBench.newPr    many-seps  avgt    5  1111.462 ± 21.815  ns/op
```

</details>

**Comment by strelec** [comment]
Thanks for your input and review @som-snytt!

This is good to commit now. With these nits:

1. There is mixed indent and non-indent code. Does not look pretty. Personally, I will never be able to get used to indent code, but up to reviewers.
2. There are unrelated scaladoc edits.
3. Useful first index location is not used in the surrogate codepath path, but that is okay, because this code is only there for correctness. Nobody actually splits strings by a surrogate.

Please review.

File: library/src/scala/collection/StringOps.scala [hunk 5]
```diff
@@ -857,19 +859,19 @@ final class StringOps(private val s: String) extends AnyVal { self =>
...
    *  the substrings terminated by the start of the string, the end of the
-   *  string or the separator character, excluding empty trailing substrings
+   *  string or the separator character, excluding empty trailing substrings.
    *
    *  If the separator character is a surrogate character, only split on
-   *  matching surrogate characters if they are not part of a surrogate pair
+   *  matching surrogate characters if they are not part of a surrogate pair.
    *
-   *  The behaviour follows, and is implemented in terms of <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html#split(java.lang.String)">String.split(re: String)</a>
+   *  This method is specified by <a href="https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/String.html#split(java.lang.String)">String.split(regex: String)</a>.
```
Comment (tanishiking) [inline library/src/scala/collection/StringOps.scala]: `This method is specified by` feels like too strong a claim to me, in a sense that `String.split(ch: Char)` does not accept regex special characters.

I think it would be more good to either use the original `The behaviour follows, and is implemented in terms of`, or add a note such as `except that this method does not accept regex special characters.` ?

What do you think? @som-snytt

File: library/src/scala/collection/StringOps.scala [hunk 6]
```diff
@@ -898,7 +900,60 @@ final class StringOps(private val s: String) extends AnyVal { self =>
    *
    *  @return an array of strings computed by splitting this string around occurrences of the separator character
    */
-  def split(separator: Char): Array[String] = s.split(escape(separator))
+  def split(separator: Char): Array[String] = {
+    val first = s.indexOf(separator)
+    if (first == -1)
+      return Array(s)
```
Comment (tanishiking) [inline library/src/scala/collection/StringOps.scala:918]: Ah, I was convinced that we shouldn't make every other case slower, but if we use `first` anyway, we're good to go 😄

File: library/test/scala/collection/StringOpsTest.scala [hunk 7]
```diff
@@ -112,4 +112,37 @@ class StringOpsTest {
...
+
+    for
+      len <- 0 to 5
+      s <- allStringsOfLength(len)
+      ch <- separators
+    do
+      assertSameElements(s.split(s"$ch"), s.split(ch))
+  }
+
+  @Test def `split assertions`: Unit = {
```
Comment (tanishiking) [inline library/test/scala/collection/StringOpsTest.scala]: Can you add more test cases like:

- empty string case (`"".split('a')` => `Array("")`)
- no matching case (`"abc".split('x')`)
- adjacent delimiters in the middle (`"a:::b".split(':')`)
- regex matcher delimiters like `$`, `.`, `|` (they should behave differently than `String.split(String)`)

?

File: library/test/scala/collection/StringOpsTest.scala [hunk 8]
```diff
@@ -112,4 +112,37 @@ class StringOpsTest {
...
+    def check(s: String, c: Char) = assertSameElements(s.split(c.toString), s.split(c))
```
Comment (tanishiking) [inline library/test/scala/collection/StringOpsTest.scala]: For testing regex matcher delimiters, and if we use `String.split(String)` as an oracle,
we might want to use `s.split(Pattern.quote(c.toString))` as an test oracle.

Comment (tanishiking) [reply library/test/scala/collection/StringOpsTest.scala]: What do you think about test against `expected`, in addition to test oracle? It would be more readable :) 

```suggestion
    def check(s: String, c: Char, expected: Array[String]) =
      val actual = s.split(c)
      assertSameElements(s.split(Pattern.quote(c.toString)), actual)
      assertSameEleemnts(expected, actual)
```

**Review by tanishiking — COMMENTED** [review]
The new algorithm looks good to me 💯 , added some comments on tests

File: library/src/scala/collection/StringOps.scala (hunk 5 above)
Comment (sjrd) [reply library/src/scala/collection/StringOps.scala]: I would write:

> A call to `s.split(c)` behaves as `s.split(java.util.regex.Pattern.quote(c.toString()))`, and therefore specified by <link-to-String.split>

File: library/src/scala/collection/StringOps.scala [hunk 9]
```diff
@@ -898,7 +900,60 @@ final class StringOps(private val s: String) extends AnyVal { self =>
...
+
+    var end = s.length
+    while (end > 0 && s(end - 1) == separator)
```
Comment (sjrd) [inline library/src/scala/collection/StringOps.scala]:
```suggestion
    while (end > first && s(end - 1) == separator)
```

File: library/src/scala/collection/StringOps.scala (hunk 6 above)
Comment (strelec) [reply library/src/scala/collection/StringOps.scala:918]: Exactly. Starting out with indexOf is especially beneficial, because it can be intrinsifed into SIMD instructions by the JDK.

File: library/test/scala/collection/StringOpsTest.scala (hunk 7 above)
Comment (strelec) [reply library/test/scala/collection/StringOpsTest.scala]: I will just delete this test.

Why?

1. Examples are given in the Scaladoc of the method already.
2. All of these possibilities you mention are already included in the above test with `allStringsOfLength`.

File: library/src/scala/collection/StringOps.scala (hunk 5 above)
Comment (strelec) [reply library/src/scala/collection/StringOps.scala]: I agree. Moreover, this is so important, I must be mentioned in the very first line of the scaladoc. Currently, the first line just says "Splits this string around the separator character.", which is misleading.

**Comment by strelec** [comment]
I have done a few last fixes, see the comments.

**Review by tanishiking — CHANGES_REQUESTED** [review]
Sorry for the delayed reply
I still believe we should have explicit unit tests that covers the cases I commented in https://github.com/scala/scala3/pull/26036#discussion_r3280753826

While the existing smoke test is kinda valuable,  unit tests should include example cases where the expected results are visible. Otherwise, when a regression happens, it's hard to tell what behavior was actually broken.

> I will just delete this test.
> Why?
> Examples are given in the Scaladoc of the method already.
> All of these possibilities you mention are already included in the above test with allStringsOfLength.

- scaladoc example isn't an enough reason not to have explicit test.
- Those properties are not tested in your smoke test, the input string and separators are too narrow. we should test
  - regex special char separators like `.`, `$`, and `\\`.
  - whitespace/control: `\n`, `\t`, `\u0000`

File: library/src/scala/collection/StringOps.scala [hunk 10]
```diff
@@ -898,7 +897,60 @@ final class StringOps(private val s: String) extends AnyVal { self =>
...
+          } else {
+            builder += s.substring(start, idx)
+            collect(idx + 1)
+          }
+        }
+        collect(first + 1)
+        builder.result()
+      }
+    } else {
+      val builder = Array.newBuilder[String]
```
Comment (He-Pin) [inline library/src/scala/collection/StringOps.scala:946]: set a size hint

Comment (tanishiking) [reply library/src/scala/collection/StringOps.scala:946]: No. To provide an accurate size hint, we would first need an extra pass to count the separators. This might help in cases with many separators to trigger repeated buffer growth, but I don’t think it would improve performance otherwise.

Comment (He-Pin) [reply library/src/scala/collection/StringOps.scala:946]: I have a question about this PR, the current PR leverages the String#indexOf to leverage the Vector support in JVM, but the current implementation of ScalaJS and ScalaNative lacks the support with SIMD. I'm not sure if @strelec wants to submit a PR in the scala-js and scala-native repos too.

**Comment by strelec** [comment]
I have vibe-coded some tests now. Other than that, no AI is used.

File: library/src/scala/collection/StringOps.scala (hunk 10 above)
Comment (strelec) [reply library/src/scala/collection/StringOps.scala:946]: Technically, we can supply the size hint of 2, because at this point, it is likely the array is going to be of size 2 or more.

But calling setHint(2) might be counter-productive in terms of performance.

Also, is Array.newBuilder even efficient? I have checked the source code and it's pretty complex. Does JDK optimize it to just an array and simple methods, or is each append quite expensive?

Comment (sjrd) [reply library/src/scala/collection/StringOps.scala:946]: A `setHint` here would be counter-productive, I agree. `setHint` is almost always a bad idea if we don't know *exactly* how many elements we will produce.

File: library/test/scala/collection/StringOpsTest.scala [hunk 11]
```diff
@@ -112,4 +112,72 @@ class StringOpsTest {
...
+
+    // --- LOW SURROGATE SPLITS ---
+    
+    // Basic isolated low surrogate split
+    assertSameElements(s"a${low}b".split(low), Array("a", "b"))
+
+    // Should skip the low surrogate because it belongs to a valid pair
+    assertSameElements(s"a${pair}b".split(low), Array(s"a${pair}b"))
+
+    // CRITICAL EDGE CASE: Over-trimming recovery
```
Comment (sjrd) [inline library/test/scala/collection/StringOpsTest.scala]: Typical AI over-exaggeration:
```suggestion
    // Edge case: Over-trimming recovery
```

**Review by sjrd — APPROVED** [review]

File: library/src/scala/collection/StringOps.scala (hunk 10 above)
Comment (strelec) [reply library/src/scala/collection/StringOps.scala:946]: Good point.

And besides, we don't even have any statistical data how to output array lengths are distributed, so we cannot guess a fixed number that would benefit most while hurting only few.

**Review by tanishiking — APPROVED** [review]
LGTM, thanks a lot for working through this!

**Comment by strelec** [comment]
> LGTM, thanks a lot for working through this!

Of course! Do we know what Scala release is this going to be included in?

**Comment by sjrd** [comment]
Not in 3.9.0, but yes in 3.9.1.

It's unlikely that there will be a 3.8.5, so not in 3.8.x either. Earlier versions depend on the Scala 2.13 library, so if you want it there, you'll have to submit a backporting PR to scala/scala.
